File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic/BoundLateMemberAccess.cs
using System.Diagnostics;
using Microsoft.CodeAnalysis.VisualBasic.Symbols;

namespace Microsoft.CodeAnalysis.VisualBasic
{
	internal sealed class BoundLateMemberAccess : BoundExpression
	{
		private readonly string _NameOpt;

		private readonly TypeSymbol _ContainerTypeOpt;

		private readonly BoundExpression _ReceiverOpt;

		private readonly BoundTypeArguments _TypeArgumentsOpt;

		private readonly LateBoundAccessKind _AccessKind;

		public string NameOpt => _NameOpt;

		public TypeSymbol ContainerTypeOpt => _ContainerTypeOpt;

		public BoundExpression ReceiverOpt => _ReceiverOpt;

		public BoundTypeArguments TypeArgumentsOpt => _TypeArgumentsOpt;

		public LateBoundAccessKind AccessKind => _AccessKind;

		public BoundLateMemberAccess SetAccessKind(LateBoundAccessKind newAccessKind)
		{
			return Update(NameOpt, ContainerTypeOpt, ReceiverOpt, TypeArgumentsOpt, newAccessKind, base.Type);
		}

		public BoundLateMemberAccess(SyntaxNode syntax, string nameOpt, TypeSymbol containerTypeOpt, BoundExpression receiverOpt, BoundTypeArguments typeArgumentsOpt, LateBoundAccessKind accessKind, TypeSymbol type, bool hasErrors = false)
			: base(BoundKind.LateMemberAccess, syntax, type, hasErrors || BoundNodeExtensions.NonNullAndHasErrors(receiverOpt) || BoundNodeExtensions.NonNullAndHasErrors(typeArgumentsOpt))
		{
			_NameOpt = nameOpt;
			_ContainerTypeOpt = containerTypeOpt;
			_ReceiverOpt = receiverOpt;
			_TypeArgumentsOpt = typeArgumentsOpt;
			_AccessKind = accessKind;
		}

		[DebuggerStepThrough]
		public override BoundNode Accept(BoundTreeVisitor visitor)
		{
			return visitor.VisitLateMemberAccess(this);
		}

		public BoundLateMemberAccess Update(string nameOpt, TypeSymbol containerTypeOpt, BoundExpression receiverOpt, BoundTypeArguments typeArgumentsOpt, LateBoundAccessKind accessKind, TypeSymbol type)
		{
			if ((object)nameOpt != NameOpt || (object)containerTypeOpt != ContainerTypeOpt || receiverOpt != ReceiverOpt || typeArgumentsOpt != TypeArgumentsOpt || accessKind != AccessKind || (object)type != base.Type)
			{
				BoundLateMemberAccess boundLateMemberAccess = new BoundLateMemberAccess(base.Syntax, nameOpt, containerTypeOpt, receiverOpt, typeArgumentsOpt, accessKind, type, base.HasErrors);
				boundLateMemberAccess.CopyAttributes(this);
				return boundLateMemberAccess;
			}
			return this;
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/BoundLateMemberAccess.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
Imports System
Imports System.Diagnostics

Namespace Microsoft.CodeAnalysis.VisualBasic
	Friend NotInheritable Class BoundLateMemberAccess
		Inherits BoundExpression
		Private ReadOnly _NameOpt As String

		Private ReadOnly _ContainerTypeOpt As TypeSymbol

		Private ReadOnly _ReceiverOpt As BoundExpression

		Private ReadOnly _TypeArgumentsOpt As BoundTypeArguments

		Private ReadOnly _AccessKind As LateBoundAccessKind

		Public ReadOnly Property AccessKind As LateBoundAccessKind
			Get
				Return Me._AccessKind
			End Get
		End Property

		Public ReadOnly Property ContainerTypeOpt As TypeSymbol
			Get
				Return Me._ContainerTypeOpt
			End Get
		End Property

		Public ReadOnly Property NameOpt As String
			Get
				Return Me._NameOpt
			End Get
		End Property

		Public ReadOnly Property ReceiverOpt As BoundExpression
			Get
				Return Me._ReceiverOpt
			End Get
		End Property

		Public ReadOnly Property TypeArgumentsOpt As BoundTypeArguments
			Get
				Return Me._TypeArgumentsOpt
			End Get
		End Property

		Public Sub New(ByVal syntax As SyntaxNode, ByVal nameOpt As String, ByVal containerTypeOpt As TypeSymbol, ByVal receiverOpt As BoundExpression, ByVal typeArgumentsOpt As BoundTypeArguments, ByVal accessKind As LateBoundAccessKind, ByVal type As TypeSymbol, Optional ByVal hasErrors As Boolean = False)
			MyBase.New(BoundKind.LateMemberAccess, syntax, type, If(hasErrors OrElse receiverOpt.NonNullAndHasErrors(), True, typeArgumentsOpt.NonNullAndHasErrors()))
			Me._NameOpt = nameOpt
			Me._ContainerTypeOpt = containerTypeOpt
			Me._ReceiverOpt = receiverOpt
			Me._TypeArgumentsOpt = typeArgumentsOpt
			Me._AccessKind = accessKind
		End Sub

		<DebuggerStepThrough>
		Public Overrides Function Accept(ByVal visitor As BoundTreeVisitor) As BoundNode
			Return visitor.VisitLateMemberAccess(Me)
		End Function

		Public Function SetAccessKind(ByVal newAccessKind As LateBoundAccessKind) As BoundLateMemberAccess
			Return Me.Update(Me.NameOpt, Me.ContainerTypeOpt, Me.ReceiverOpt, Me.TypeArgumentsOpt, newAccessKind, MyBase.Type)
		End Function

		Public Function Update(ByVal nameOpt As String, ByVal containerTypeOpt As TypeSymbol, ByVal receiverOpt As BoundExpression, ByVal typeArgumentsOpt As BoundTypeArguments, ByVal accessKind As LateBoundAccessKind, ByVal type As TypeSymbol) As Microsoft.CodeAnalysis.VisualBasic.BoundLateMemberAccess
			Dim boundLateMemberAccess As Microsoft.CodeAnalysis.VisualBasic.BoundLateMemberAccess
			If (CObj(nameOpt) <> CObj(Me.NameOpt) OrElse CObj(containerTypeOpt) <> CObj(Me.ContainerTypeOpt) OrElse receiverOpt <> Me.ReceiverOpt OrElse typeArgumentsOpt <> Me.TypeArgumentsOpt OrElse accessKind <> Me.AccessKind OrElse CObj(type) <> CObj(MyBase.Type)) Then
				Dim boundLateMemberAccess1 As Microsoft.CodeAnalysis.VisualBasic.BoundLateMemberAccess = New Microsoft.CodeAnalysis.VisualBasic.BoundLateMemberAccess(MyBase.Syntax, nameOpt, containerTypeOpt, receiverOpt, typeArgumentsOpt, accessKind, type, MyBase.HasErrors)
				boundLateMemberAccess1.CopyAttributes(Me)
				boundLateMemberAccess = boundLateMemberAccess1
			Else
				boundLateMemberAccess = Me
			End If
			Return boundLateMemberAccess
		End Function
	End Class
End Namespace